src/note.rs

```diff
@@ -27,8 +167,132 @@ impl From<u8> for Note {
     }
 }
 
-impl Into<u8> for Note {
-    fn into(self) -> u8 {
-        self.0
+impl From<Note> for u8 {
+    fn from(value: Note) -> Self {
+        value.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(u8)]
+pub enum NoteName {
+    C,
+    Cs,
+    D,
+    Ds,
+    E,
+    F,
+    Fs,
+    G,
+    Gs,
+    A,
+    As,
+    B,
+}
+
+impl From<NoteName> for u8 {
+    fn from(value: NoteName) -> Self {
+        value as u8
+    }
+}
+
+#[allow(non_upper_case_globals)]
+impl NoteName {
+    pub const Db: NoteName = NoteName::Cs;
+    pub const Eb: NoteName = NoteName::Ds;
+    pub const Gb: NoteName = NoteName::Fs;
+    pub const Ab: NoteName = NoteName::Gs;
+    pub const Bb: NoteName = NoteName::As;
+}
+
+impl From<Note> for (NoteName, i8) {
+    fn from(note: Note) -> Self {
+        let octave: i8 = note.0 as i8 / 12 - 1;
+        let note_name = match note.0 % 12 {
+            name if name == NoteName::C as u8 => NoteName::C,
+            name if name == NoteName::Cs as u8 => NoteName::Cs,
+            name if name == NoteName::D as u8 => NoteName::D,
+            name if name == NoteName::Ds as u8 => NoteName::Ds,
+            name if name == NoteName::E as u8 => NoteName::E,
+            name if name == NoteName::F as u8 => NoteName::F,
+            name if name == NoteName::Fs as u8 => NoteName::Fs,
+            name if name == NoteName::G as u8 => NoteName::G,
+            name if name == NoteName::Gs as u8 => NoteName::Gs,
+            name if name == NoteName::A as u8 => NoteName::A,
+            name if name == NoteName::As as u8 => NoteName::As,
+            name if name == NoteName::B as u8 => NoteName::B,
+            _ => panic!("This code should be unreachable"), // Out of bound values should not occur because of the %12
+        };
+
+        (note_name, octave)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum MidiConversionError {
+    OutOfBoundsError,
+}
+
+impl TryFrom<(NoteName, i8)> for Note {
```

Comment:
Maybe we could provide an additional From implementation that just returns the note clamped between min and max. We're doing this everywhere else so maybe provide this here too.

Reply:
it looks like rustc won't let you impl both a `TryFrom` and `From` (or `TryInto` and `Into`) to -> from the same type

I'm fine to ditch the `Try` if you think that is better

Reply:
Can't we just do an `impl From<(NoteName, i8)> for Note` next to the TryFrom? Then you can just choose if you want to do a `.try_into()` or `.into()`. Having to chain them seems to be useless overhead. And it also bloats the Error type if you dont want to use the clamped value.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f393729cb3505d66f4a97d3ca76b10e

```
conflicting implementation in crate `core`:
          - impl<T, U> TryFrom<U> for T
            where U: Into<T>;
```

https://doc.rust-lang.org/std/convert/trait.TryFrom.html#impl-TryFrom%3CU%3E-for-T

looks like `TryFrom` is automatically implemented if the type implements `Into` which is automatic if you implement `From` so we get a conflict.

Reply:
I guess we could do something like this?
```rust
impl TryFrom<(NoteName, i8)> for Note {
    type Error = NoteOutOfBoundsError;

    fn try_from(value: (NoteName, i8)) -> Result<Self, Self::Error> {
        let (name, octave) = value;
        if octave < -2i8 {
            Err(NoteOutOfBoundsError { clamped: Note::MIN })?
        } else if octave > 8 {
            Err(NoteOutOfBoundsError { clamped: Note::MAX })?
        }

        let number = name as u8 + (octave + 2) as u8 * 12;

        if number > 127 {
            Err(NoteOutOfBoundsError { clamped: Note::MAX })
        } else {
            Ok(Note::new(number))
        }
    }
}

impl From<(i8, NoteName)> for Note {
    fn from(value: (i8, NoteName)) -> Note {
        let (octave, name) = value;
        if octave < -2i8 {
            Note::MIN
        } else if octave > 8 {
            Note::MAX
        } else {
            let number = name as u8 + (octave + 2) as u8 * 12;

            if number > 127 {
                Note::MAX
            } else {
                Note::new(number)
            }
        }
    }
}

```

Reply:
We could also just make these extra constructors on Note,
Or even remove the NoteName + octave stuff completely as its not really central to Midi anyway.

Reply:
I'm fine with either one of those.. I wouldn't likely use any of the note name stuff myself, but i figure some might find some value in it.